### PR TITLE
Add intel_amx backend for Radix Attention

### PR DESCRIPTION
**Pull request by yanbing-j**
## Motivation

This PR is to add intel_amx attention backend for CPU, and update extend_attention/decode_attention kernel implementation, and add the corresponding UTs test_extend.py/test_decode.py as well.

**Note that**:

- If device is set to `cpu` and attention backend is not set, then it will use intel_amx attention backend if hardware supports it.
- We fuse `set_kv_buffer` in `decode` attention, in order to reduce the overhead.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by mingfeima** [comment]
@yanbing-j  we also need kernel level test cases: `decode_attention` and `extend_attention`. Will help us debug future optimizations.

**Comment by mingfeima** [comment]
you can cherry-pick commits from our developing branch [cpu_opt_ww11](https://github.com/mingfeima/sglang/tree/cpu_opt_ww11) if necessary, as this will keep original commit messages.

**Comment by mingfeima** [comment]
use cherry-pick, don't directly replace files from our working branch.

File: test/srt/cpu/test_decode.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        configs = [
+            (2, 16, 16, 64, 64),
+            (2, 16, 1, 16, 16),
+            (2, 32, 8, 33, 55),
+            (2, 16, 1, 64, 64),
+            (2, 64, 1, 13, 13),
+            (2, 128, 1, 80, 80),
+            (2, 128, 2, 512, 512),
+            (1, 16, 1, 576, 512),
+            # (1, 16, 16, 576, 512), # nan not greater than 0.99
```
Comment (mingfeima) [inline test/srt/cpu/test_decode.py]: this indicates that the set kv buffer function in `MHA` path went wrong.

File: test/srt/cpu/test_extend.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            b_req_idx,
+            b_seq_len,
+            b_seq_len_extend,
+            b_start_loc_extend,
+            max_len_extend,
+            sm_scale,
+            logit_cap)
+
+        print(o_ref)
+        print(o_extend)
```
Comment (mingfeima) [inline test/srt/cpu/test_extend.py]:
```suggestion
```

File: test/srt/cpu/test_decode.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        causal=False,
+    ):
+        # [num_tokens, num_heads, head_size] -> [num_heads, num_tokens, head_size]
+        query = query.movedim(0, query.dim() - 2)
+
+        start_q, start_kv = 0, 0
+        for seq_idx in range(seq_lens.shape[0]):
+            # TODO: this loop process a sequence per iter, this is inefficient.
+            # Need optimize the performance later.
+
```
Comment (mingfeima) [inline test/srt/cpu/test_decode.py]:
```suggestion
```

File: sgl-kernel/csrc/cpu/torch_extension_cpu.cpp [hunk 4]
```diff
@@ -47,17 +47,10 @@ std::tuple<at::Tensor, at::Tensor> biased_grouped_topk_cpu(
...
-    at::Tensor& attn_logits,
-    at::Tensor& req_to_token,
-    at::Tensor& req_pool_indices,
-    at::Tensor& seq_lens,
-    double sm_scale,
-    double logit_cap);
+void decode_attention_cpu(at::Tensor& query, at::Tensor& k_cache, at::Tensor& v_cache, at::Tensor& output,
+    at::Tensor& key, at::Tensor& value, at::Tensor& loc, at::Tensor& attn_logits,
+    at::Tensor& req_to_token, at::Tensor& req_pool_indices, at::Tensor& seq_lens,
+    double sm_scale, double logit_cap);
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/torch_extension_cpu.cpp]: the original formatting is correct.

File: sgl-kernel/csrc/cpu/gemm.h [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <ATen/native/CPUBlas.h>
+#include "common.h"
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/gemm.h]:
```suggestion
```

Comment (mingfeima) [inline sgl-kernel/csrc/cpu/gemm.h]:
```suggestion
```

File: sgl-kernel/csrc/cpu/gemm.cpp [hunk 6]
```diff
@@ -487,7 +487,7 @@ at::Tensor weight_packed_linear(at::Tensor& mat1, at::Tensor& mat2, std::optiona
   const float* bias_data = nullptr;
   if (has_bias) {
     CHECK_EQ(bias.value().size(0), N);
-    bias_data = bias.value().data_ptr<float>();
+    bias_data = bias.value().to(at::kFloat).data_ptr<float>();
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/gemm.cpp]: we don't need `to.(at::kFloat)` here. our cpu_opt_ww11 branch doesn't have this.

if the model has non-float32 bias, we need to convert when packing the weights to vnni formats in python APIs not the C++ ones.

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
@yanbing-j this PR covers too much extra scope: it has tensor parallel related staff and also the MoE layers change. I expect this PR only covers that:
* intel amx attention backend -  both decode and extend, python API enabling
* kernel level unit tests - just make sure that `test_decode.py` and `test_extend.py` can pass

You don't have to the model level evaluation, just kernel level tests are enough.

So please remove:
https://github.com/sgl-project/sglang/pull/6143/commits/61af296c7aa49f9cdf491fec00b12e67efe135c4 
https://github.com/sgl-project/sglang/pull/6143/commits/9c1a570d85c0d808f1f2780db7bfb5fc2f637398

File: sgl-kernel/csrc/cpu/gemm.h (hunk 5 above)
Comment (yanbing-j) [reply sgl-kernel/csrc/cpu/gemm.h]: Need this due to the build failure `In file included from /home/yanbingj/project/sglang/sgl-kernel/csrc/cpu/gemm.cpp:1:
/home/yanbingj/project/sglang/sgl-kernel/csrc/cpu/gemm.h:55:5: error: ‘Tensor’ in namespace ‘at’ does not name a type
   55 | at::Tensor convert_weight_packed(at::Tensor& weight);
      |     ^~~~~~
`

File: sgl-kernel/csrc/cpu/gemm.cpp (hunk 6 above)
Comment (yanbing-j) [reply sgl-kernel/csrc/cpu/gemm.cpp]: It is introduced by https://github.com/mingfeima/sglang/commit/6f73059a93758b49f428b5739f188e6bb9bd27cd, I remove this line change.

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            forward_batch.token_to_kv_pool.set_kv_buffer(
+                layer, forward_batch.out_cache_loc, k, v
+            )
+
+        _, max_extend_len = self.forward_metadata
+
+        self.extend_attention_fwd(
+            q.view(-1, layer.tp_q_head_num, layer.qk_head_dim),
+            k.contiguous(),
+            v.contiguous(),
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py]: i think underlying C++ kernels should be able to handle non-contiguous input with the last dimension, please check!
if so, remove `contiguous()` here and also update test cases

File: python/sglang/srt/models/deepseek_v2.py [hunk 8]
```diff
@@ -510,13 +510,14 @@ def __init__(
         if rope_scaling:
             rope_scaling["rope_type"] = "deepseek_yarn"
 
-        self.rotary_emb = get_rope(
+        self.rotary_emb = get_rope_wrapper(
```
Comment (mingfeima) [inline python/sglang/srt/models/deepseek_v2.py]: do we need to update this file?

File: test/srt/cpu/test_extend.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        assert seq_lens.shape[0] == extend_seq_lens.shape[0]
+
+        # [num_tokens, num_heads, head_size] -> [num_heads, num_tokens, head_size]
+        query = query.movedim(0, query.dim() - 2)
+
+        start_q, start_kv = 0, 0
+        for seq_idx in range(seq_lens.shape[0]):
+            # TODO: this loop process a sequence per iter, this is inefficient.
+            # Need optimize the performance later.
+
```
Comment (mingfeima) [inline test/srt/cpu/test_extend.py]:
```suggestion

```

**Review by mingfeima — CHANGES_REQUESTED** [review]

**Comment by mingfeima** [comment]
put a note on intel_amx behavior
* device set cpu, attn not set - use intel_amx if hardware supports it.

File: python/sglang/srt/layers/attention/intel_amx_backend.py (hunk 7 above)
Comment (yanbing-j) [reply python/sglang/srt/layers/attention/intel_amx_backend.py]: Yes, non-contiguous inputs of k, v, k_buffer, v_buffer is accepted by extend_attention_cpu. UTs are updated.

File: python/sglang/srt/models/deepseek_v2.py (hunk 8 above)
Comment (yanbing-j) [reply python/sglang/srt/models/deepseek_v2.py]: This is cherry-pick from the commit of enabling intel_amx, used for model running, will this break the upstream?

Comment (mingfeima) [reply python/sglang/srt/models/deepseek_v2.py]: let's remove it in this PR. And later on, when the individual OPs have been enabled in sglang, we fix model level changes in just one sperate PR.

Comment (yanbing-j) [reply python/sglang/srt/models/deepseek_v2.py]: Done.

**Comment by mingfeima** [comment]
@yanbing-j rebase as https://github.com/sgl-project/sglang/pull/6115 been landed.

File: python/sglang/srt/managers/schedule_batch.py [hunk 10]
```diff
@@ -1226,7 +1226,10 @@ def prepare_for_extend(self):
         self.extend_input_logprob_token_ids = extend_input_logprob_token_ids
 
         # Write to req_to_token_pool
-        if global_server_args_dict["attention_backend"] != "torch_native":
+        if not global_server_args_dict["attention_backend"] in [
+            "torch_native",
+            "intel_amx",
+        ]:
```
Comment (mingfeima) [inline python/sglang/srt/managers/schedule_batch.py]:
```suggestion
        if global_server_args_dict.get("attention_backend") not in ("torch_native", "intel_amx"):
```
more readable.

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 11]
```diff
@@ -336,7 +336,10 @@ def init_new(
             ret.extend_prefix_lens = torch.tensor(
                 batch.extend_prefix_lens, dtype=torch.int32
             ).to(device, non_blocking=True)
-            if model_runner.server_args.attention_backend != "torch_native":
+            if (
+                model_runner.server_args.attention_backend != "torch_native"
+                and model_runner.server_args.attention_backend != "intel_amx"
+            ):
```
Comment (mingfeima) [inline python/sglang/srt/model_executor/forward_batch_info.py]:
```suggestion
            if model_runner.server_args.attention_backend not in ("torch_native", "intel_amx"):
```
more readable

File: python/sglang/srt/server_args.py [hunk 12]
```diff
@@ -292,6 +292,11 @@ def __post_init__(self):
             self.sampling_backend = "pytorch"
 
         # Set kernel backends
+        if self.device == "cpu":
+            if self.attention_backend is None:
+                self.attention_backend = "intel_amx"
```
Comment (mingfeima) [inline python/sglang/srt/server_args.py:303]: what if device is `cpu` and it is not an AMX CPU, say AMD? Do we have logic to convert self.attention_backend back to torch_native ?

**Review by mingfeima — CHANGES_REQUESTED** [review]
Almost done here, just some minor issues to address.

File: python/sglang/srt/server_args.py (hunk 12 above)
Comment (yanbing-j) [reply python/sglang/srt/server_args.py:303]: Will add this in another PR.

**Review by mingfeima — APPROVED** [review]

**Comment by yanbing-j** [comment]
Hi @zhyncs , could you please help review this PR? Thanks!

File: python/sglang/srt/cpu_utils.py
Comment (Alcanderian) [inline python/sglang/srt/cpu_utils.py:1]: please merge this file into python/sglang/srt/utils.py

File: python/sglang/srt/managers/schedule_batch.py [hunk 13]
```diff
@@ -1226,7 +1226,10 @@ def prepare_for_extend(self):
         self.extend_input_logprob_token_ids = extend_input_logprob_token_ids
 
         # Write to req_to_token_pool
-        if global_server_args_dict["attention_backend"] != "torch_native":
+        if global_server_args_dict.get("attention_backend") not in (
```
Comment (Alcanderian) [inline python/sglang/srt/managers/schedule_batch.py:1253]: Here is a bit ambiguous and I think this logic could be convert to a new arguments `--disable-triton` in server args. And whats more there maybe more devices that do not support triton(for example ascend npu) @zhyncs @merrymercy

Comment (mingfeima) [reply python/sglang/srt/managers/schedule_batch.py:1253]: probably we can add an attribute named maybe `support_triton` in the attention module abstraction. And get this attr here from `schedule_batch.py`.

this will avoid the user level argument turning to be too complexed. User doesn't need to be aware of the information whether attention backends like `torch_native` or `intel_amx` will support triton or not. The developers who implements these should handle this.

one advantage of doing this is that for future hardwares, ascend npu, ppu, etc. they only need to instantiate `support_triton` when they implement radix attention and don't have to modify `schedule_batch.py`.

basically, what we wan to have on the cpu backend is something like:
* user will ONLY need to set the device to be `cpu`
* attention backend choose to use `intel_amx` which is optimized C++ and assembly kernels for intel Xeon with amx support
* when the hardware does not support amx (older x86 servers for example), it fall back to torch_native backend which should be equally functional but of low performance.

Comment (mingfeima) [reply python/sglang/srt/managers/schedule_batch.py:1253]: @yanbing-j

**Comment by zhyncs** [comment]
Hi @yanbing-j @mingfeima Please consider using two separate PRs for this change: one for the sgl-kernel related changes and another for the integration of the Python part. After the sgl-kernel PR, can we also add Intel CPU CI unit tests to ensure that the Python integration part functions properly? Thank you! BTW I can help setup GitHub runners.

**Comment by mingfeima** [comment]
> Hi @yanbing-j @mingfeima Please consider using two separate PRs for this change: one for the sgl-kernel related changes and another for the integration of the Python part. After the sgl-kernel PR, can we also add Intel CPU CI unit tests to ensure that the Python integration part functions properly? Thank you! BTW I can help setup GitHub runners.

sure, working in progress!

**Comment by mingfeima** [comment]
replaced with
https://github.com/sgl-project/sglang/pull/6405
https://github.com/sgl-project/sglang/pull/6408
